### PR TITLE
SALTO-4050 fix more namespace errors and edit log to show full filePr…

### PR DESCRIPTION
**Pull request by benisraeladi**
Fix errors in namespace when API name is complex (has a `.` in the full name). Also edit log so that when the namespace prefix isn't added on purpose the log shows the full fileProps

---

---
_Release Notes_: _None_

---
_User Notifications_:  _None_

### CODE REVIEW

File: packages/salesforce-adapter/src/fetch.ts [hunk 1]
```diff
@@ -175,6 +176,12 @@ const getFullName = (obj: FileProperties, addNamespacePrefixToFullName?: boolean
   if (!obj.namespacePrefix) {
     return obj.fullName
   }
+
+  // Managed package names should exactly match obj.namespacePrefix
+  if (obj.fullName === obj.namespacePrefix) {
```
Comment (tamtamirr) [inline packages/salesforce-adapter/src/fetch.ts:181]: Is this code relevant for metadata instances of type **InstalledPackage** only?
If so, wouldn't that be more accurate?
```suggestion
  if (obj.type === INSTALLED_PACKAGE_METADATA) {
```

File: packages/salesforce-adapter/src/fetch.ts [hunk 2]
```diff
@@ -188,15 +195,21 @@ const getFullName = (obj: FileProperties, addNamespacePrefixToFullName?: boolean
...
   if (addNamespacePrefixToFullName) {
-    // In some cases, obj.fullName does not contain the namespace prefix even though
-    // obj.namespacePrefix is defined. In these cases, we want to add the prefix manually
+  // In some cases, obj.fullName does not contain the namespace prefix even though
+  // obj.namespacePrefix is defined. In these cases, we want to add the prefix manually
+    if (obj.fullName.includes(API_NAME_SEPARATOR)) {
+    // Case for API name with an object <Object>.X where it should be <Object>.<namespace>__X
+      const [nameWithoutNamespace, nameParts] = obj.fullName.split(API_NAME_SEPARATOR).reverse()
+      return `${makeArray(nameParts).reverse().join(API_NAME_SEPARATOR)}${API_NAME_SEPARATOR}${namePrefix}${nameWithoutNamespace}`
+    }
```
Comment (tamtamirr) [inline packages/salesforce-adapter/src/fetch.ts:209]: Simplify as we discussed f2f :)

File: packages/salesforce-adapter/test/fetch_installed_packages.test.ts [hunk 3]
```diff
@@ -92,6 +92,32 @@ describe('Test fetching installed package metadata', () => {
       })
     })
   })
+  describe('Test fetching objects with complicated API names', () => {
+    describe('addNamespacePrefixToFullName is true', () => {
```
Comment (tamtamirr) [inline packages/salesforce-adapter/test/fetch_installed_packages.test.ts:96]:
```suggestion
    describe('when addNamespacePrefixToFullName is true', () => {
```

File: packages/salesforce-adapter/test/fetch_installed_packages.test.ts [hunk 4]
```diff
@@ -92,6 +92,32 @@ describe('Test fetching installed package metadata', () => {
...
+        const expectedFullName = 'Account.Test__TestObject'
+        const addNamespacePrefixToFullName = true
+        const fileProp = mockFileProperties({ fullName: fullNameFromList, type: 'Account', namespacePrefix: 'Test' })
+        const instance = await fetch({ fileProp, mockType: mockTypes.PermissionSet, addNamespacePrefixToFullName })
+
+        expect(instance).toBeDefined()
+        expect(instance?.value).toHaveProperty('fullName', expectedFullName)
+      })
+    })
+    describe('addNamespacePrefixToFullName is false', () => {
```
Comment (tamtamirr) [inline packages/salesforce-adapter/test/fetch_installed_packages.test.ts:108]:
```suggestion
    describe('when addNamespacePrefixToFullName is false', () => {
```

File: packages/salesforce-adapter/test/fetch_installed_packages.test.ts [hunk 5]
```diff
@@ -92,6 +92,32 @@ describe('Test fetching installed package metadata', () => {
...
+      it('should not add prefix to the object\'s name', async () => {
+        const fullNameFromList = 'TestPermissionSet'
+        const expectedFullName = 'TestPermissionSet'
+        const addNamespacePrefixToFullName = false
+        const fileProp = mockFileProperties({ fullName: fullNameFromList, type: 'PermissionSet', namespacePrefix: 'Test' })
+        const instance = await fetch({ fileProp, mockType: mockTypes.PermissionSet, addNamespacePrefixToFullName })
+
+        expect(instance).toBeDefined()
+        expect(instance?.value).toHaveProperty('fullName', expectedFullName)
```
Comment (tamtamirr) [inline packages/salesforce-adapter/test/fetch_installed_packages.test.ts:117]: Nit: Consider the following instead
```suggestion
        expect(instance).toEqual(expect.objectContaining({
          value: expect.objectContaining({ [INSTANCE_FULL_NAME_FIELD]: expectedFullName })
        }))
```

File: packages/salesforce-adapter/test/fetch_installed_packages.test.ts [hunk 6]
```diff
@@ -92,6 +92,32 @@ describe('Test fetching installed package metadata', () => {
...
+      })
+    })
+    describe('addNamespacePrefixToFullName is false', () => {
+      it('should not add prefix to the object\'s name', async () => {
+        const fullNameFromList = 'TestPermissionSet'
+        const expectedFullName = 'TestPermissionSet'
```
Comment (tamtamirr) [inline packages/salesforce-adapter/test/fetch_installed_packages.test.ts:111]: Nit: Is this two vars really necessary? I would just inline them were they are used. I'm not sure it adds readability to the test.

File: packages/salesforce-adapter/src/fetch.ts (hunk 1 above)
Comment (tamtamirr) [inline packages/salesforce-adapter/src/fetch.ts:181]: Please add a UT for this scenario.

**Review by tamtamirr — COMMENTED** [review]
Looking good 👑 Added some comments

File: packages/salesforce-adapter/src/fetch.ts [hunk 7]
```diff
@@ -175,6 +177,12 @@ const getFullName = (obj: FileProperties, addNamespacePrefixToFullName?: boolean
   if (!obj.namespacePrefix) {
     return obj.fullName
   }
+
+  // Managed package names should exactly match obj.namespacePrefix
```
Comment (tamtamirr) [inline packages/salesforce-adapter/src/fetch.ts:181]:
```suggestion
  // Instances of type InstalledPackage fullNames should never include the namespace prefix
```

File: packages/salesforce-adapter/src/fetch.ts [hunk 8]
```diff
@@ -188,15 +196,20 @@ const getFullName = (obj: FileProperties, addNamespacePrefixToFullName?: boolean
...
   }
+
   if (obj.fullName.startsWith(namePrefix)) {
     return obj.fullName
   }
   if (addNamespacePrefixToFullName) {
-    // In some cases, obj.fullName does not contain the namespace prefix even though
-    // obj.namespacePrefix is defined. In these cases, we want to add the prefix manually
+  // In some cases, obj.fullName does not contain the namespace prefix even though
+  // obj.namespacePrefix is defined. In these cases, we want to add the prefix manually
```
Comment (tamtamirr) [inline packages/salesforce-adapter/src/fetch.ts:205]: indentation

File: packages/salesforce-adapter/src/fetch.ts [hunk 9]
```diff
@@ -188,15 +196,20 @@ const getFullName = (obj: FileProperties, addNamespacePrefixToFullName?: boolean
...
+    if (obj.fullName.includes(API_NAME_SEPARATOR)) {
+    // Case for API name with an object <Object>.X where it should be <Object>.<namespace>__X
+      return obj.fullName.replace(API_NAME_SEPARATOR, `${API_NAME_SEPARATOR}${namePrefix}`)
```
Comment (tamtamirr) [inline packages/salesforce-adapter/src/fetch.ts:208]: Nit:
While this is correct, this logic seems vague for me, and since this value is **EXTREMELY** important I'd prefer to be explicit here.
```suggestion
      [parentName, instanceName] = obj.fullName.split(API_NAME_SEPARATOR)
      return `${parentName}.${namePrefix}${instanceName}`
```

File: packages/salesforce-adapter/test/fetch_installed_packages.test.ts [hunk 10]
```diff
@@ -92,6 +93,45 @@ describe('Test fetching installed package metadata', () => {
...
+    describe('when addNamespacePrefixToFullName is true', () => {
+      it('should add prefix to the object\'s name correctly', async () => {
+        const fullNameFromList = 'Account.TestObject'
+        const expectedFullName = 'Account.Test__TestObject'
+        const addNamespacePrefixToFullName = true
+        const fileProp = mockFileProperties({ fullName: fullNameFromList, type: 'Account', namespacePrefix: 'Test' })
+        const instance = await fetch({ fileProp, mockType: mockTypes.PermissionSet, addNamespacePrefixToFullName })
+
+        expect(instance).toBeDefined()
+        expect(instance?.value).toHaveProperty('fullName', expectedFullName)
```
Comment (tamtamirr) [inline packages/salesforce-adapter/test/fetch_installed_packages.test.ts:106]: I've seen you replaced the other tests with my previous suggestion, probably missed this one..? It's better to be aligned in all the tests :) (Those that were part of this PR at-least, I wouldn't touch code that is unrelated to this PR)

**Review by tamtamirr — APPROVED** [review]
LGTM 🥳 Added some comments.
Approving to save cycles.

File: packages/salesforce-adapter/src/fetch.ts (hunk 9 above)
Comment (ori-moisis) [reply packages/salesforce-adapter/src/fetch.ts:208]: I really think we should not make this change "blindly"
we should stick with logging first to make sure we don't ruin other scenarios that currently work

Comment (benisraeladi) [reply packages/salesforce-adapter/src/fetch.ts]: @ori-moisis There won't be any change unless the flag is off. Until then, there will only be logging on the failed scenario. Do you mean to split the flag up so that each sub-scenario is logged?
